app/src/main/java/com/azuresamples/msalnativeauthandroidkotlinsampleapp/MFAFragment.kt

```diff
@@ -0,0 +1,236 @@
+        private enum class STATUS { SignedIn, SignedOut }
+    }
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        _binding = FragmentEmailPasswordBinding.inflate(inflater, container, false)
+        val view = binding.root
+
+        (activity as? AppCompatActivity)?.supportActionBar?.title = getString(R.string.title_email_otp_mfa)
+
+        authClient = AuthClient.getAuthClient()
+
+        init()
+
+        return view
+    }
+
+    override fun onResume() {
+        super.onResume()
+        getStateAndUpdateUI()
+    }
+
+    private fun init() {
+        binding.signUp.visibility = View.GONE
+
+        initializeButtonListeners()
+    }
+
+    private fun initializeButtonListeners() {
+        binding.signIn.setOnClickListener {
+            signIn()
+        }
+
+        binding.signOut.setOnClickListener {
+            signOut()
+        }
+    }
+
+    private fun getStateAndUpdateUI() {
+        CoroutineScope(Dispatchers.Main).launch {
+            val accountResult = authClient.getCurrentAccount()
+            when (accountResult) {
+                is GetAccountResult.AccountFound -> {
+                    displaySignedInState(accountResult.resultValue)
+                }
+                is GetAccountResult.NoAccountFound -> {
+                    displaySignedOutState()
+                }
+                is GetAccountError -> {
+                    displayDialog(getString(R.string.msal_exception_title), accountResult.errorMessage)
+                }
+            }
+        }
+    }
+
+    private fun signIn() {
+        CoroutineScope(Dispatchers.Main).launch {
+            val email = binding.emailText.text.toString()
+            val password = CharArray(binding.passwordText.length())
+            binding.passwordText.text?.getChars(0, binding.passwordText.length(), password, 0)
+
+            val actionResult: SignInResult = authClient.signIn(
+                username = email,
+                password = password
+            )
+            binding.passwordText.text?.clear()
+            StringUtil.overwriteWithNull(password)
+
+            when (actionResult) {
+                is SignInResult.Complete -> {
+                    Toast.makeText(
+                        requireContext(),
+                        getString(R.string.sign_in_successful_message),
+                        Toast.LENGTH_SHORT
+                    ).show()
+                    displaySignedInState(accountState = actionResult.resultValue)
+                }
+                is SignInResult.MFARequired -> {
+                    val awaitingMFAState = actionResult.nextState
+                    val checkDefaultResult = awaitingMFAState.sendChallenge()
+                    if (checkDefaultResult is MFARequiredResult.VerificationRequired) {
+                        navigateToMFARequired(nextState = checkDefaultResult.nextState)
```

Comment:
nit, but important from the point of view of teaching developers how this flow works:
- at this point we're further than "MFA required"; we're in "verification required". I like the approach of naming some handling function "MFARequired", because it shows that we're now entering the MFA path, but this is not the right place. Or to rephrase: this exact method should be called something like `navigateToMFAVerification`, and a previous method/navigation should be named `handleMFARequired`). For example: you could consider showing a dialog when `MFARequired` is received that explains to the user they must perform MFA, and when the user clicks "proceed" to call `sendChallenge()`. Was this discussed as an option during the sample app proposal review?

Reply:
Done.